### ISSUE TITLE
get_max_score_handicap is slowing down test suite

### ISSUE THREAD

**Issue by TomHall2020** [jatkinson1000/archeryutils#75]
Running the test suite locally on my machine takes around 4-6 seconds, but it is obvious from watching it that there are a few particularly slow tests that I have to wait to finish one by one. `pytest --durations` shows that the slowest tests are all related to `get_max_score_handicap`, in fact 12 tests account for about 2.5s out of the total execution time of 3.5s here.

```sh
❯ python -W ignore -m pytest --durations 20
=============================================================== test session starts ===============================================================
platform darwin -- Python 3.11.5, pytest-7.4.4, pluggy-1.3.0
rootdir: /Users/Tom/dev/forks/archeryutils
plugins: mock-3.12.0
collected 369 items

archeryutils/classifications/tests/test_agb_field.py ...............................                                                        [  8%]
archeryutils/classifications/tests/test_agb_indoor.py .......................................                                               [ 18%]
archeryutils/classifications/tests/test_agb_old_indoor.py .....................                                                             [ 24%]
archeryutils/classifications/tests/test_agb_outdoor.py ..................................................                                   [ 38%]
archeryutils/classifications/tests/test_classification_utils.py ..........                                                                  [ 40%]
archeryutils/handicaps/tests/test_handicap_tables.py .......................                                                                [ 47%]
archeryutils/handicaps/tests/test_handicaps.py ............................................................................................ [ 72%]
.......                                                                                                                                     [ 73%]
archeryutils/tests/test_constants.py ..............                                                                                         [ 77%]
archeryutils/tests/test_rounds.py ...................................                                                                       [ 87%]
archeryutils/tests/test_targets.py ...............................................                                                          [100%]

============================================================== slowest 20 durations ===============================================================
0.32s call     archeryutils/handicaps/tests/test_handicaps.py::TestHandicapFromScore::test_maximum_score[AGB-testround0-720-11]
0.32s call     archeryutils/handicaps/tests/test_handicaps.py::TestHandicapFromScore::test_get_max_score_handicap[testround0-AGB-True-11]
0.26s call     archeryutils/handicaps/tests/test_handicaps.py::TestHandicapFromScore::test_maximum_score[AA-testround1-720-107]
0.26s call     archeryutils/handicaps/tests/test_handicaps.py::TestHandicapFromScore::test_get_max_score_handicap[testround1-AA-True-107]
0.21s call     archeryutils/handicaps/tests/test_handicaps.py::TestHandicapFromScore::test_get_max_score_handicap[testround2-AGB-False-9.89]
0.21s call     archeryutils/handicaps/tests/test_handicaps.py::TestHandicapFromScore::test_maximum_score[AGB-testround2-864-9]
0.18s call     archeryutils/handicaps/tests/test_handicaps.py::TestHandicapFromScore::test_get_max_score_handicap[testround3-AGBold-True-6]
0.18s call     archeryutils/handicaps/tests/test_handicaps.py::TestHandicapFromScore::test_maximum_score[AGBold-testround3-864-6]
0.17s call     archeryutils/handicaps/tests/test_handicaps.py::TestHandicapFromScore::test_get_max_score_handicap[testround4-AGB-True-3]
0.17s call     archeryutils/handicaps/tests/test_handicaps.py::TestHandicapFromScore::test_maximum_score[AGB-testround4-300-3]
0.13s call     archeryutils/handicaps/tests/test_handicaps.py::TestHandicapFromScore::test_get_max_score_handicap[testround5-AA-False-118.38]
0.13s call     archeryutils/handicaps/tests/test_handicaps.py::TestHandicapFromScore::test_maximum_score[AA-testround5-300-119]
0.02s call     archeryutils/handicaps/tests/test_handicap_tables.py::TestHandicapTable::test_check_print_table_inputs[hcs0-False-expected0]

(7 durations < 0.005s hidden.  Use -vv to show these durations.)
=============================================================== 369 passed in 3.47s ===============================================================
```

I profiled the `get_max_score_handicap` function with cProfile and found that it was taking about 6500 iterations to find the handicap for max score on a York round, which is a relatively generous case as that is an AGB handicap of -9.8. Easier rounds will take even more iterations to solve, and are relatively more likely to be requested (maximum scores on a York are pretty unlikely for most...)

```
❯ python -W ignore scratch/debug_slow_max_score.py
         247587 function calls in 0.353 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.353    0.353 <string>:1(<module>)
     6515    0.001    0.000    0.001    0.000 fromnumeric.py:2172(_sum_dispatcher)
     6515    0.010    0.000    0.047    0.000 fromnumeric.py:2177(sum)
     6515    0.012    0.000    0.036    0.000 fromnumeric.py:71(_wrapreduction)
     6515    0.006    0.000    0.006    0.000 fromnumeric.py:72(<dictcomp>)
    19545    0.037    0.000    0.037    0.000 handicap_equations.py:174(sigma_t)
    19545    0.011    0.000    0.048    0.000 handicap_equations.py:287(sigma_r)
    19545    0.073    0.000    0.257    0.000 handicap_equations.py:338(arrow_score)
    97725    0.109    0.000    0.109    0.000 handicap_equations.py:421(<genexpr>)
     6515    0.008    0.000    0.286    0.000 handicap_equations.py:517(score_for_passes)
     6515    0.017    0.000    0.274    0.000 handicap_equations.py:570(<listcomp>)
     6515    0.012    0.000    0.347    0.000 handicap_equations.py:580(score_for_round)
        1    0.006    0.006    0.353    0.353 handicap_functions.py:150(get_max_score_handicap)
        3    0.000    0.000    0.000    0.000 rounds.py:143(max_score)
        1    0.000    0.000    0.000    0.000 rounds.py:236(max_score)
        4    0.000    0.000    0.000    0.000 rounds.py:245(<genexpr>)
        3    0.000    0.000    0.000    0.000 targets.py:176(max_score)
        1    0.000    0.000    0.000    0.000 {built-in method _warnings.warn}
        1    0.000    0.000    0.353    0.353 {built-in method builtins.exec}
     6515    0.002    0.000    0.002    0.000 {built-in method builtins.isinstance}
    19546    0.027    0.000    0.136    0.000 {built-in method builtins.sum}
     6515    0.005    0.000    0.005    0.000 {built-in method numpy.array}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
     6515    0.001    0.000    0.001    0.000 {method 'items' of 'dict' objects}
     6515    0.017    0.000    0.017    0.000 {method 'reduce' of 'numpy.ufunc' objects}


round_=Round('York')
Result of get_max_score_handicap(york, 'AGB')
===================
-9.869999999998377
```

Fundamentally the issue is just taking too many iterations at a small step size of `0.01` on the handicap scale. That accuracy is only needed when we are very close to a handicap that will produce a sub-maximum score, everything before then is wasted. I think I was able to fix this and have a PR ready that implements it.

**Comment by jatkinson1000** [comment]
Yep, I noted this when writing that there were optimisations to be made in the search algorithm, but wasn't a priority.

Happy for you to open a MR.

I am getting close to uploading the refactored Handicaps code which is why I have been holding off on other aspects for a while. Code seems done, now tidying typing and writing docstrings etc. Will also need to update docs, but I will do that after opening the MR for you to look at.

**Comment by TomHall2020** [comment]
Results of tests and profiling with patch in place:

```
❯ python -W ignore -m pytest --durations 20
=============================================================== test session starts ===============================================================
platform darwin -- Python 3.11.5, pytest-7.4.4, pluggy-1.3.0
rootdir: /Users/Tom/dev/forks/archeryutils
plugins: mock-3.12.0
collected 369 items

archeryutils/classifications/tests/test_agb_field.py ...............................                                                        [  8%]
archeryutils/classifications/tests/test_agb_indoor.py .......................................                                               [ 18%]
archeryutils/classifications/tests/test_agb_old_indoor.py .....................                                                             [ 24%]
archeryutils/classifications/tests/test_agb_outdoor.py ..................................................                                   [ 38%]
archeryutils/classifications/tests/test_classification_utils.py ..........                                                                  [ 40%]
archeryutils/handicaps/tests/test_handicap_tables.py .......................                                                                [ 47%]
archeryutils/handicaps/tests/test_handicaps.py ............................................................................................ [ 72%]
.......                                                                                                                                     [ 73%]
archeryutils/tests/test_constants.py ..............                                                                                         [ 77%]
archeryutils/tests/test_rounds.py ...................................                                                                       [ 87%]
archeryutils/tests/test_targets.py ...............................................                                                          [100%]

============================================================== slowest 20 durations ===============================================================
0.01s call     archeryutils/handicaps/tests/test_handicap_tables.py::TestHandicapTable::test_check_print_table_inputs[hcs0-False-expected0]
0.01s call     archeryutils/handicaps/tests/test_handicaps.py::TestHandicapFromScore::test_maximum_score[AGB-testround0-720-11]
0.01s call     archeryutils/handicaps/tests/test_handicaps.py::TestHandicapFromScore::test_get_max_score_handicap[testround0-AGB-True-11]
0.01s call     archeryutils/handicaps/tests/test_handicaps.py::TestHandicapFromScore::test_maximum_score[AGB-testround2-864-9]
0.01s call     archeryutils/handicaps/tests/test_handicaps.py::TestHandicapFromScore::test_get_max_score_handicap[testround2-AGB-False-9.89]

(15 durations < 0.005s hidden.  Use -vv to show these durations.)
=============================================================== 369 passed in 0.89s ===============================================================


❯ python -W ignore scratch/debug_slow_max_score.py
         3605 function calls in 0.008 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.008    0.008 <string>:1(<module>)
       92    0.000    0.000    0.000    0.000 fromnumeric.py:2172(_sum_dispatcher)
       92    0.000    0.000    0.001    0.000 fromnumeric.py:2177(sum)
       92    0.000    0.000    0.001    0.000 fromnumeric.py:71(_wrapreduction)
       92    0.000    0.000    0.000    0.000 fromnumeric.py:72(<dictcomp>)
      276    0.001    0.000    0.001    0.000 handicap_equations.py:174(sigma_t)
      276    0.000    0.000    0.001    0.000 handicap_equations.py:287(sigma_r)
      276    0.002    0.000    0.006    0.000 handicap_equations.py:338(arrow_score)
     1380    0.003    0.000    0.003    0.000 handicap_equations.py:421(<genexpr>)
       92    0.000    0.000    0.007    0.000 handicap_equations.py:517(score_for_passes)
       92    0.000    0.000    0.006    0.000 handicap_equations.py:570(<listcomp>)
       92    0.000    0.000    0.008    0.000 handicap_equations.py:580(score_for_round)
        1    0.000    0.000    0.008    0.008 handicap_functions.py:150(get_max_score_handicap)
       92    0.000    0.000    0.008    0.000 handicap_functions.py:197(check_score)
        3    0.000    0.000    0.000    0.000 rounds.py:143(max_score)
        1    0.000    0.000    0.000    0.000 rounds.py:236(max_score)
        4    0.000    0.000    0.000    0.000 rounds.py:245(<genexpr>)
        3    0.000    0.000    0.000    0.000 targets.py:176(max_score)
        1    0.000    0.000    0.000    0.000 {built-in method _warnings.warn}
        1    0.000    0.000    0.008    0.008 {built-in method builtins.exec}
       92    0.000    0.000    0.000    0.000 {built-in method builtins.isinstance}
      277    0.001    0.000    0.003    0.000 {built-in method builtins.sum}
       92    0.000    0.000    0.000    0.000 {built-in method numpy.array}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
       92    0.000    0.000    0.000    0.000 {method 'items' of 'dict' objects}
       92    0.000    0.000    0.000    0.000 {method 'reduce' of 'numpy.ufunc' objects}


round_=Round('York')
Result of get_max_score_handicap(round_, 'AGB')
===================
-9.870000000000005
```

Down to 92 evaluations of `score_for_round`, the result is the same within floating point precision and can get the full test suite done in about 1s which feels way nicer.

**Comment by jatkinson1000** [comment]
This was closed by #76
